packages/arb-node-core/cmdhelp/wallet.go

```diff
@@ -27,75 +28,122 @@ import (
+	"github.com/rs/zerolog/log"
 	"golang.org/x/crypto/ssh/terminal"
 
 	"github.com/offchainlabs/arbitrum/packages/arb-util/configuration"
 )
 
+var logger = log.With().Caller().Stack().Str("component", "cmdhelp").Logger()
+
 // GetKeystore returns a transaction authorization based on an existing ethereum
 // keystore located in validatorFolder/wallets or creates one if it does not
 // exist. It accepts a password using the "password" command line argument or
 // via an interactive prompt. It also sets the gas price of the auth via an
 // optional "gasprice" arguement.
 func GetKeystore(
-	validatorFolder string,
-	wallet *configuration.Wallet,
-	gasPrice float64,
+	config *configuration.Config,
+	walletConfig *configuration.Wallet,
 	chainId *big.Int,
 ) (*bind.TransactOpts, func([]byte) ([]byte, error), error) {
-	ks := keystore.NewKeyStore(
-		filepath.Join(validatorFolder, "wallets"),
-		keystore.StandardScryptN,
-		keystore.StandardScryptP,
-	)
-
 	var account accounts.Account
-	if len(ks.Accounts()) > 0 {
-		account = ks.Accounts()[0]
-	}
+	var signer = func(data []byte) ([]byte, error) { return nil, errors.New("undefined signer") }
+	var auth *bind.TransactOpts
 
-	if ks.Unlock(account, wallet.Password) != nil {
-		if len(ks.Accounts()) == 0 {
-			fmt.Print("Enter new account password: ")
-		} else {
-			fmt.Print("Enter account password: ")
+	if len(walletConfig.FireblocksPrivateKey) != 0 {
+		fromAddress := ethcommon.HexToAddress(config.Fireblocks.SourceAddress)
+		auth = &bind.TransactOpts{
+			From: fromAddress,
+			Signer: func(address ethcommon.Address, tx *types.Transaction) (*types.Transaction, error) {
+				if address != fromAddress {
+					logger.Error().Hex("currentaddress", address.Bytes()).Hex("expectedaddress", fromAddress.Bytes()).Msg("incorrect from address provided")
+					return nil, bind.ErrNotAuthorized
+				}
+				// Just return original unsigned transaction because fireblocks will handle signing
+				return tx, nil
+			},
 		}
 
-		bytePassword, err := terminal.ReadPassword(int(syscall.Stdin))
+		signer = func(data []byte) ([]byte, error) {
+			// Fireblocks cannot sign arbitrary data
+			return make([]byte, 32), nil
+		}
+	} else if len(config.Wallet.PrivateKey) != 0 {
+		privateKey, err := crypto.HexToECDSA(config.Wallet.PrivateKey)
+		if err != nil {
+			return nil, nil, err
+		}
+		auth, err = bind.NewKeyedTransactorWithChainID(privateKey, chainId)
 		if err != nil {
 			return nil, nil, err
 		}
-		passphrase := string(bytePassword)
 
-		passphrase = strings.TrimSpace(passphrase)
+		signer = func(data []byte) ([]byte, error) {
+			return crypto.Sign(data, privateKey)
+		}
+	} else {
+		ks := keystore.NewKeyStore(
+			filepath.Join(config.Persistent.Chain, "wallets"),
+			keystore.StandardScryptN,
+			keystore.StandardScryptP,
+		)
+
+		if len(ks.Accounts()) > 0 {
+			account = ks.Accounts()[0]
+		}
+
+		if ks.Unlock(account, walletConfig.Password) != nil {
+			if len(walletConfig.Password) == 0 {
```

Comment:
I've previously used an empty password with a wallet. I'm not sure if this was supported in the old version either but something to consider.

Reply:
The old version always prompted for password if wallet failed to open.  The new version only prompts for password if wallet failed to open and no password (or empty password) is specified.

Empty password can still be used by just pressing return at prompt, or providing a pregenerated wallet.  I think that is a reasonable compromise